lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,234 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from opaque_keys.edx.locator import BlockUsageLocator
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from hashlib import md5
+import json
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=32, primary_key=True)
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return self.hashed
+
+    @classmethod
+    def create(cls, blocks):
+        """
+        Creates a new VisibleBlocks model. Argument 'blocks' should be an array of BlockRecords.
+        """
+        # Start by sorting the blocks to ensure equality regardless of how the blocks are ordered
+        sorted_blocks = sorted(
```

Comment:
(Preface: I really don't know the correct answer to this)

What does that gain us though? Is it useful to fail on an isinstance check at this point instead of an AttributeError when accessing members? Are there problems with allowing anything that defines `block_key`, `max_score`, `weight`, and `to_dict()` to be serialized in this manner?